masquerade-core/src/main/java/com/flipkart/masquerade/processor/ReferenceMapProcessor.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2017 Flipkart Internet, pvt ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.flipkart.masquerade.processor;
+
+import com.flipkart.masquerade.Configuration;
+import com.flipkart.masquerade.rule.Rule;
+import com.squareup.javapoet.*;
+
+import javax.lang.model.element.Modifier;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.flipkart.masquerade.util.Helper.getRuleInterface;
+
+/**
+ * Processor that adds and initializes a Map field mapping Class to Mask Object.
+ * <p />
+ * Created by shrey.garg on 12/05/17.
+ */
+public class ReferenceMapProcessor {
+    private final Configuration configuration;
+    private final TypeSpec.Builder cloakBuilder;
+
+    /**
+     * @param configuration Configuration for the current processing cycle
+     * @param cloakBuilder Entry class under construction for the cycle
+     */
+    public ReferenceMapProcessor(Configuration configuration, TypeSpec.Builder cloakBuilder) {
+        this.configuration = configuration;
+        this.cloakBuilder = cloakBuilder;
+    }
+
+    /**
+     * Adds a private field and initializes it as well.
+     *
+     * @param rule The rule for which the Map will be generated
+     */
+    public void addMap(Rule rule) {
+        ParameterizedTypeName mapType = ParameterizedTypeName.get(ClassName.get(Map.class), ClassName.get(String.class), getRuleInterface(configuration, rule));
+        cloakBuilder.addField(FieldSpec.builder(mapType, rule.getName(), Modifier.PRIVATE).initializer("new $T<>()", HashMap.class).build());
```

Comment:
For a complex use case, this will have a NFR concern. Lets say there hundreds of entities in an application and a complex multi level objects graph (representing relation between entities). If all these entities have fields to be cloaked, we have to end up traversing the entire graph and there is no other go. But for an use case where very few fields of few objects in the graph has to be cloaked, we still end up traversing the entire graph.

As such traversing the entire graph itself might not have penalty (but still needs to be measured). Penalty could come from this map containing all entities and their mask objects. At each level of graph traversal, we end up doing a look up in this map (which might not be o(1) always). There was an earlier comment that this map might have to support a list of mask objects as its value, in that case, penalty could still go up a bit.